smithy-model/src/main/java/software/amazon/smithy/model/node/DefaultNodeDeserializers.java

```diff
@@ -0,0 +1,594 @@
+                }
+            };
+        } catch (NoSuchMethodException e) {
+            // Don't fail, and instead try the next deserializer.
+            return null;
+        }
+    };
+
+    // Creates enums by checking each enum variant and detecting if the result of
+    // calling toString on the variant matches the given string.
+    // Mimic's Jackson's behavior when using READ_ENUMS_USING_TO_STRING
+    // See https://github.com/FasterXML/jackson-databind/wiki/Deserialization-Features
+    private static final ObjectCreatorFactory ENUM_CREATOR = (nodeType, target) -> {
+        if (nodeType != NodeType.STRING || !Enum.class.isAssignableFrom(target)) {
+            return null;
+        }
+
+        return (node, targetType, parameterizedType, pointer, mapper) -> {
+            String name = node.expectStringNode().getValue();
+            for (Object constant : targetType.getEnumConstants()) {
+                if (constant.toString().equals(name)) {
+                    return constant;
+                }
+            }
+
+            // Give an error message with suggestions.
+            List<String> names = new ArrayList<>();
+            for (Object constant : targetType.getEnumConstants()) {
+                names.add(constant.toString());
+            }
+
+            throw NodeDeserializationException.fromContext(targetType, pointer, node, null,
+                    "Expected one of the following enum strings: " + names);
+        };
+    };
+
+    private interface FromStringClassFactory {
+        Object create(String input) throws Exception;
+    }
+
+    // These types have special, built-in handling.
+    // This mirrors the simpler behaviors allowed in Jackson.
+    // See https://github.com/FasterXML/jackson-databind/blob/ab583fb2319ee33ef6b548b720afec84265d40a7/src/main/java/com/fasterxml/jackson/databind/deser/std/FromStringDeserializer.java
+    private static final Map<Class, FromStringClassFactory> FROM_STRING_CLASSES = MapUtils.of(
+            URL.class, URL::new,
+            URI.class, URI::new,
+            Pattern.class, Pattern::compile,
+            Path.class, Paths::get,
+            File.class, File::new
+    );
+
+    private static final ObjectCreatorFactory FROM_STRING = (nodeType, target) -> {
+        if (nodeType != NodeType.STRING || !FROM_STRING_CLASSES.containsKey(target)) {
+            return null;
+        }
+
+        FromStringClassFactory factory = FROM_STRING_CLASSES.get(target);
+        return (node, targetType, parameterizedType, pointer, mapper) -> {
+            String value = node.expectStringNode().getValue();
+            try {
+                return factory.create(value);
+            } catch (Exception e) {
+                throw NodeDeserializationException.fromContext(
+                        targetType, pointer, node, e, e.getMessage());
+            }
+        };
+    };
+
+    private static final List<ObjectCreatorFactory> DEFAULT_FACTORIES = ListUtils.of(
```

Comment:
Minor: Would be good to have a comment here indicating that these are priority ordered, and what determined the priority.